### PR TITLE
Add HL7 struct storing HL7 data as nested maps

### PR DESCRIPTION
**Pull request by DarkMarmot**
This now has an HL7 module to generate map data from HL7 text to store in an HL7 struct. It also includes functions to query these maps using a `~p` sigil (an HL7 path). This adopts standards and structures that are incompatible with `HL7.Query`. My goal here is to eventually replicate the functionality found with these maps. In the interim, you can move between the two as the data structures are compact but lossless in round robins between this and HL7.Message.

### CODE REVIEW

File: benchmark.exs [hunk 1]
```diff
@@ -4,14 +4,15 @@ alias HL7.Examples
...
+list = msg |> Message.to_list()
 
-list = Examples.wikipedia_sample_hl7() |> HL7.Message.to_list() |> List.flatten() |> Enum.take(10)
 Benchee.run(%{
-  "raw" => fn -> msg |> Message.raw() end,
-  "new" => fn -> msg |> Message.new(%{accept_latin1: true}) end,
-  "new-copy" => fn -> msg |> Message.new(%{copy: true}) end,
-  "new-alt" => fn -> alt |> Message.new() end,
-  "round-trip" => fn -> msg |> Message.new() |> to_string() end
+#  "raw" => fn -> msg |> Message.raw() end,
```
Comment (starbelly) [inline benchmark.exs:10]: Looks completely dead, is that so?

Comment (DarkMarmot) [reply benchmark.exs]: cleaned up!

File: lib/hl7/hpath_parser.ex [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  defparsec(
+    :parse,
+    segment_header
+    |> optional(field)
+    |> optional(repeat_num)
+    |> optional(component_part)
+    |> optional(bang)
+    |> eos(),
+    export_metadata: true
```
Comment (adkron) [inline lib/hl7/hpath_parser.ex:54]: `export_metadata` is mostly needed for testing and I think it can slow things down.

**Comment by adkron** [comment]
As a thought, should we do `~hl{PID-19}`. Often there is a lowercase version, and then the upper adds some interpolation or something like that.

I also like when I don't have to use the shift key. 😉

**Comment by DarkMarmot** [comment]
> As a thought, should we do `~hl{PID-19}`. Often there is a lowercase version, and then the upper adds some interpolation or something like that.
> 
> I also like when I don't have to use the shift key. 😉

I would vast prefer not shifting as well, but I think multichar sigils only support uppercase right now?

File: lib/hl7/hpath_parser.ex (hunk 2 above)
Comment (DarkMarmot) [reply lib/hl7/hpath_parser.ex:54]: ahh cool, i'll nix it thanks!

File: benchmark.exs [hunk 3]
```diff
@@ -56,11 +57,11 @@ Benchee.run(%{
...
-# new              51.93 K       19.26 μs    ±28.73%       17.37 μs       38.47 μs
-# new-copy         47.77 K       20.93 μs    ±29.84%       19.21 μs       39.15 μs
-# round-trip       20.32 K       49.21 μs    ±18.52%       47.62 μs       80.37 μs
-# new-alt          15.50 K       64.53 μs    ±12.32%       62.72 μs       97.16 μs
+# add Maps output for use with the new HPath sigil
 
+# raw             137.72 K
+# new              43.00 K - 3.20x slower +15.99 μs
+# new-copy         38.38 K - 3.59x slower +18.80 μs
+# maps             32.75 K - 4.21x slower +23.27 μs
```
Comment (adkron) [inline benchmark.exs:65]: Do we make up for the slower creation with the searching?

Do you have a benchmark on searching?

Comment (DarkMarmot) [reply benchmark.exs:65]: I haven't done one. You're welcome to make some more if you want? I think the ergonomic benefits make it worthwhile even for the extra speed hit. There's likely some number of lookups that enumerate lists the old way we'd have to do before we'd hit break even. Probably a lot. But we're doing a lot now when you consider every event handler we have...

File: lib/hl7/hpath.ex [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    {:ok, data, _, _, _, _} = HL7.HPathParser.parse(term)
+
+    path_map =
+      %__MODULE__{}
+      |> Map.merge(Map.new(data, fn {k, v} -> {k, hd(v)} end))
+
+    path_map
+    |> then(&Map.put(&1, :indices, get_indices(&1)))
+    |> Macro.escape()
+  end
```
Comment (adkron) [inline lib/hl7/hpath.ex:77]: We need a non-sigil way to build these for people who need to be able to do this at runtime.

Comment (DarkMarmot) [reply lib/hl7/hpath.ex]: Good point! added, still need test....

Comment (DarkMarmot) [reply lib/hl7/hpath.ex:94]: and doc tested!

File: lib/hl7.ex [hunk 5]
```diff
@@ -2,14 +2,201 @@ defmodule HL7 do
...
+          optional(pos_integer) => hl7_map_data() | String.t(),
+          e: non_neg_integer()
+        }
+
+  @type t() :: %__MODULE__{segments: [segment()]}
+
+  @type parsed_hl7_segments :: t() | [segment()]
+  @type parsed_hl7 :: t() | segment() | [segment()] | hl7_map_data()
+
+  alias HL7.Path
```
Comment (adkron) [inline lib/hl7.ex:28]: Should we add the path sigil here so people can start to only use the `HL7` module as the public face of the library?

File: lib/hl7.ex [hunk 6]
```diff
@@ -2,14 +2,201 @@ defmodule HL7 do
...
+    text |> HL7.Message.to_list() |> parse!()
+  end
+
+  def parse!(%HL7.Message{} = message) do
+    message |> HL7.Message.to_list() |> parse!()
+  end
+
+  @spec parse(String.t()) :: {:ok, t()} | {:error, HL7.InvalidMessage.t()}
+  def parse(text) when is_binary(text) do
+    case HL7.Message.new(text) do
```
Comment (adkron) [inline lib/hl7.ex:54]: It is strange to me that HL7.Message is deprecated, but we are using it here in the new function.

File: benchmark.exs (hunk 3 above)
Comment (adkron) [reply benchmark.exs:65]:
```elixir
list_message = HL7.Message.new(msg)
map_message = HL7.parse!(list_message)

path = ~p{PID-18}
deep_path = ~p{PID-11.4}

Benchee.run(%{
  "list" => fn -> HL7.Query.find_first(list_message, path) end,
  "map" => fn -> HL7.get(map_message, path) end
})
```

```
Name           ips        average  deviation         median         99th %
list        3.07 M      326.00 ns ±11292.07%         208 ns         375 ns
map         2.69 M      372.17 ns  ±7779.30%         292 ns         458 ns

Comparison: 
list        3.07 M
map         2.69 M - 1.14x slower +46.17 ns

Name                ips        average  deviation         median         99th %
deep list        2.60 M      383.99 ns  ±9353.22%         250 ns         458 ns
deep map         2.49 M      402.21 ns  ±6922.81%         292 ns         500 ns

Comparison: 
deep list        2.60 M
deep map         2.49 M - 1.05x slower +18.21 ns
```

The map appears to be slower

Comment (DarkMarmot) [reply benchmark.exs:65]: To note, slightly slower except for perhaps things like PV1-42 or such. Biggest benefit is for seeing index values in the maps.

File: lib/hl7.ex [hunk 7]
```diff
@@ -2,14 +2,201 @@ defmodule HL7 do
...
+  alias HL7.Path
+
+  @doc ~S"""
+  Creates an HL7 struct from HL7 data (accepting text, lists or the deprecated `HL7.Message` struct).
+
+  The segment maps using integer keys corresponding to the data's HL7 positions (starting at 1).
+  Segment names are stored at position 0. To save on space and to make a cleaner presentation, empty values
+  are not included in the output. Instead, each map contains an `:e` field that notes the highest index
+  present in the source data.
+  """
```
[the remark this answers is not in the document]
Comment (DarkMarmot) [reply lib/hl7.ex:37]: I agree it's confusing and looks extraneous, so I've made it exist only internally now!

File: lib/hl7.ex (hunk 6 above)
Comment (DarkMarmot) [reply lib/hl7.ex:54]: We'll have HL7.Message for a while, as might others, and this provides an easy transition between the two for now until we deprecate/rm it in the future.

File: lib/hl7.ex (hunk 5 above)
Comment (DarkMarmot) [reply lib/hl7.ex:28]: Possibly! Will test with it.

Comment (DarkMarmot) [reply lib/hl7.ex:28]: Done!

**Review by starbelly — APPROVED** [review]
Looks good to marmot.

**Pull request merged**